### PR TITLE
SamTov update visualiser

### PR DESCRIPTION
**Pull request by SamTov**
Following additions:

* Add ZnVis visualizer
* Add example for visualizer
* Add integration test for visualizer
* Add MDSuiteProcess class for multi-process tests
* Add tests for MDSuiteProcess class
* Add test for assertDeepAlmostEqual functionality

## Important
This PR can be reviewed but not merged until ZnVis is made public. For this reason the visualizer test will currently fail. I would expect ZnVis to be public by next week.

### CODE REVIEW

**Comment by PythonFZ** [comment]
Just read it on the phone, but ZnVis is missing in the requirements.txt file. Will do a full review later.

**Comment by SamTov** [comment]
> Just read it on the phone, but ZnVis is missing in the requirements.txt file. Will do a full review later.

Yes, that was the idea behind the PR comment under the headline, important. I will update the package stuff when ZnVis is public.

**Comment by PythonFZ** [comment]
> > Just read it on the phone, but ZnVis is missing in the requirements.txt file. Will do a full review later.
> 
> Yes, that was the idea behind the PR comment under the headline, important. I will update the package stuff when ZnVis is public.

I missed that on mobile. I'd suggest converting this PR into a draft then and add add a review when ready.

**Comment by SamTov** [comment]
> > > Just read it on the phone, but ZnVis is missing in the requirements.txt file. Will do a full review later.
> > 
> > 
> > Yes, that was the idea behind the PR comment under the headline, important. I will update the package stuff when ZnVis is public.
> 
> I missed that on mobile. I'd suggest converting this PR into a draft then and add add a review when ready.

The idea is that the code passes all tests locally and is just waiting for the other package to be made public. For this reason I put it up for review as it cannot merge until that package is public however the code here will not change. It is, in this way, completely ready for review.

File: CI/integration_tests/visualizer/znvis_visualizer.py
Comment (PythonFZ) [inline CI/integration_tests/visualizer/znvis_visualizer.py:63]: Why 90 s?

File: CI/unit_tests/utils/test_testing.py
Comment (PythonFZ) [inline CI/unit_tests/utils/test_testing.py]: `assert True` should be sufficient instead of `assert True == True`.

File: CI/unit_tests/utils/test_testing.py
Comment (PythonFZ) [inline CI/unit_tests/utils/test_testing.py]: see above

File: mdsuite/calculators/spatial_distribution_function.py
Comment (PythonFZ) [inline mdsuite/calculators/spatial_distribution_function.py:254]: Why remove the `plot=True` argument? The SDF does store data in the DB https://github.com/zincware/MDSuite/blob/244a4ca81d27599e81487ea51a2d81473683e8f0/mdsuite/calculators/spatial_distribution_function.py#L246-L247 so it should have a `plot=False` option?

File: mdsuite/utils/testing.py
Comment (PythonFZ) [inline mdsuite/utils/testing.py]: This should be a super call.

File: mdsuite/utils/testing.py
Comment (PythonFZ) [inline mdsuite/utils/testing.py:67]: Can you explain this class and the associated tests to me? I don't really understand what it does tbh.

File: mdsuite/visualizer/znvis_visualizer.py
Comment (PythonFZ) [inline mdsuite/visualizer/znvis_visualizer.py]: I'd prefer using the following, because it will support refactoring:

```py
import mdsuite.data

pse = json.loads(importlib.resources.read_text(mdsuite.data, "PubChemElements_all.json"))
```

**Review by PythonFZ — CHANGES_REQUESTED** [review]
Looks good. A few comments and I can only approve the PR when the requirements.txt is added.

File: CI/integration_tests/visualizer/znvis_visualizer.py
Comment (SamTov) [reply CI/integration_tests/visualizer/znvis_visualizer.py:63]: I do explain why there is a time here in the TODO. It is a different time in the todo but the concept is the same. 

The point here is that after enough time for the test to run the process must be killed. Right now I don't include communication from the process back to the main test and so I simply wait for 90 seconds, enough time for the test to run, and then kill it. In future, we can work on a way to communicate between the threads but as long as the process test is not long this is fine.

File: mdsuite/calculators/spatial_distribution_function.py
Comment (SamTov) [reply mdsuite/calculators/spatial_distribution_function.py:254]: When plot is run it calls a method that the SDF does not have or want to have. We need to remove that or we can leave it at plot=False. As the SDF is less of a calculator and more visualisation tool I don't think it needs a plot method.

File: mdsuite/visualizer/znvis_visualizer.py
Comment (SamTov) [reply mdsuite/visualizer/znvis_visualizer.py]: yep that is a good point.

File: mdsuite/utils/testing.py
Comment (SamTov) [reply mdsuite/utils/testing.py:67]: Absolutely. If there is a better way to handle this also please let me know. 

When we want to run a test calling Open3D we need to open it. Open3D will run a separate process as it allows for callbacks but this means that it cannot be forcefully closed during it's run, it must be closed by a user. The idea behind using multiprocessing for a test is that we can open the code in a new process, let it run and then use `process.terminate()` to forcefully end it. The reason the Process is sub-classed is because we want to use it for a test and therefore need a way of collecting the exceptions or errors that were raised on the deployed process. This class collects these methods such that they can be looked at in the test. Without this functionality, if an exception is raised on the deployed process all of the tests will still pass.

File: CI/unit_tests/utils/test_testing.py
Comment (SamTov) [reply CI/unit_tests/utils/test_testing.py]: resolved in d8c6389

File: CI/unit_tests/utils/test_testing.py
Comment (SamTov) [reply CI/unit_tests/utils/test_testing.py]: resolved in d8c6389

File: mdsuite/visualizer/znvis_visualizer.py
Comment (SamTov) [reply mdsuite/visualizer/znvis_visualizer.py]: resolved in 71ff782

File: CI/unit_tests/utils/test_testing.py
Comment (PythonFZ) [inline CI/unit_tests/utils/test_testing.py]: No need for the print statement, I guess.

File: mdsuite/utils/testing.py
Comment (SamTov) [reply mdsuite/utils/testing.py]: resolved in 71ff782

Comment (SamTov) [reply mdsuite/utils/testing.py]: resolved in eb448fc

File: CI/unit_tests/utils/test_testing.py
Comment (SamTov) [reply CI/unit_tests/utils/test_testing.py]: good point, I just copied and pasted from the module and did not read that test

Comment (SamTov) [reply CI/unit_tests/utils/test_testing.py]: resolved in eb448fc

**Comment by SamTov** [comment]
I think this is resolved now.

**Review by PythonFZ — APPROVED** [review]
Maybe remove the SDF example script for now and extend the docs. But looks good!

**Review by SamTov — COMMENTED** [review]
Checking for spurious file addition. None found after conflict resolution.

**Pull request merged**
